cpp/tests/experimental/louvain_primitives.cu

```diff
@@ -0,0 +1,197 @@
+#include <experimental/graph.hpp>
+
+#include <patterns/copy_v_transform_reduce_key_aggregated_out_nbr.cuh>
+#include <patterns/transform_reduce_by_adj_matrix_row_col_key_e.cuh>
+#include <utilities/dataframe_buffer.cuh>
+#include <utilities/shuffle_comm.cuh>
+
+#include <rmm/device_uvector.hpp>
+
+template <typename graph_t>
+void update_by_delta_modularity(raft::handle_t const& handle,
+                                graph_t const& graph_view,
+                                std::vector<typename graph_t::vertex_type> cluster_h,
+                                typename graph_t::weight_type total_edge_weight,
+                                typename graph_t::weight_type resolution)
+{
+  using vertex_t = typename graph_t::vertex_type;
+  using edge_t   = typename graph_t::edge_type;
+  using weight_t = typename graph_t::weight_type;
+
+  vertex_t num_verts = graph_view.get_number_of_vertices();
+
+  rmm::device_uvector<vertex_t> cluster_v(num_verts, handle.get_stream());
+
+  raft::update_device(cluster_v.data(), cluster_h.data(), num_verts, handle.get_stream());
+
+  rmm::device_uvector<vertex_t> communities(0, handle.get_stream());
+  rmm::device_uvector<weight_t> community_in_weight_sums(0, handle.get_stream());
+
+  std::tie(communities, community_in_weight_sums) =
+    cugraph::experimental::transform_reduce_by_adj_matrix_col_key_e(
+      handle,
+      graph_view,
+      cluster_v.begin(),
+      cluster_v.begin(),
+      cluster_v.begin(),
+      [] __device__(auto r, auto c, auto w, auto rv, auto cv) {
+        printf("transform reduce (%d,%d,%g,%d,%d)\n", (int)r, (int)c, (float)w, (int)rv, (int)cv);
+        return w;
+      },
+      weight_t{0});
+
+  thrust::for_each(thrust::device,
```

Comment:
This is a debug call to display the output from the `transform_reduce_by_adj_matrix_col_key_e` call above.

The output from this is:
```
keys = 1,1,3,3,0,1,0,1,0,3,0,3
values = 1,1,1,1,1,1,1,1,1,1,1,1
```

This demonstrates the transform (the keys reflect the current clustering).  However no reduction has been done at this point.  I would have expected that the output for vertex 0 (the only one affected in this particular example) would have two keys (1 and 3) and their corresponding values would have been added (so 2 and 2).

Reply:
This sounds like a bug, there should be no repetition in keys (the output size == # unique keys). Let me take a closer look.

Reply:
The output should be keys: [0, 1, 3] values: [4, 4, 4].

and you can use https://github.com/rapidsai/raft/blob/branch-0.18/cpp/include/raft/cudart_utils.h#L252 instead of the custom for_each as a convenience utility for printing.

Reply:
I skipped a necessary reduction step in the SG path... :-(

Reply:
Isn't the objective here to do a reduce by key for each source vertex?  That's what I need.  I want the output to be:

vertex 0: { 1:2, 3:2 }
vertex 1: { 0:1, 1:1 }
vertex 2: { 0:1, 1:1 } 
vertex 3: { 0:1, 3:1 }
vertex 4: { 0:1, 3:1 }

(Using python dictionary representation of a map to describe the desired behavior).

Reply:
No, this doesn't.

This is added to compute something like Sigma_{in} in (https://en.wikipedia.org/wiki/Louvain_method); sum of all the weights of the links inside the community i. Community ID becomes a key and edge weights are values.

I think what you're looking for is ```copy_v_transform_reduce_key_aggregated_out_nbr```. This works per vertex, and for each vertex, first aggregate outgoing edges by key, and ```key_aggregated_edge_op``` is called with (src, key, sum_of_the_aggregated_edge_weights_for_the_key, src attributes, value for the key in the key, value pairs).

Reply:
I see.  I will evaluate this more closely.